KEP-2104: kube-proxy v2: reworking the proxy's architecture (#2094)
kubernetes/enhancements: keps/sig-network/20201010-kube-proxy-v2/README.md

# WORK IN PROGRESS
<!--
**Note:** When your KEP is complete, all of these comment blocks should be removed.

To get started with this template:

- [ ] **Pick a hosting SIG.**
  Make sure that the problem space is something the SIG is interested in taking
  up. KEPs should not be checked in without a sponsoring SIG.

> - [ ] **Create an issue in kubernetes/enhancements**

Reviewer: I think you didn't do this part? In particular, the KEP directory name should start with an issue number now, not with a date. (Also, you didn't copy in the `kep.yaml` from the template dir.)

Author: ahhh the issue is not the PR, ok got it and willco :)

-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

At the beginning, `kube-proxy` was designed to handle the translation of Service objects to OS-level resources.
Implementations have been userland, then iptables, and now ipvs. With the growth of the Kubernetes project, more
implementations came to life, for instance with eBPF, and often in relation to other goals (Calico to manage
the network overlay, Cilium to manage app-level security, metallb to provide an external LB for bare-metal clusters, etc).

Along this cambrian explosion of third-party software, the Service object itself received new concepts to improve the
abstraction, for instance to express topology. This, third-party implementation are expected to update and become more
complex over time, even if their core doesn't change (ie, the eBPF translation layer is not affected).

This KEP is born from the conviction that more decoupling of the Service object and the actual implementations is required,

> by introducing an intermediate, node-level abstaction provider. This abstraction is expected to be the result of applying

Reviewer: ```suggestion
by introducing an intermediate, node-level abstraction provider. This abstraction is expected to be the result of applying
```

Kubernetes' `Service` semantics and business logic to a simpler, more stable API.

> 

Reviewer: Random additional point that I don't know exactly where it belongs in the doc: having this daemon would mean that proxy implementations wouldn't need to be ported to EndpointSlice, since the Endpoints vs EndpointSlices distinction would be handled entirely within the daemon...

Author: yes, good example of the interest of the decoupling I'm trying to achieve. Same idea with adding the topology concept.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

### Goals

- provide a node-level abstraction of the cluster-wide `Service` semantics through an API
- allow easier, more stable proxy implementations that don't need updates when `Service` business logic changes
- provide a client library with minimal dependencies
- include equivalent implementations of in-project ones (userland, iptables and ipvs)
- (optional) help proxy implementations using the same subsystem (ie iptables) to cooperate more easily

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

### Non-Goals

- provide equivalent implementations of third-party ones

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. The "Design Details" section below is for the real
nitty-gritty.
-->

> Rewrite the kube-proxy to be a "localhost" gRPC API provider that will be accessible via classic TCP (`127.0.0.1:12345`)

Reviewer: "classic TCP"? Is there a "New TCP"? :slightly_smiling_face:

Author: is rephrasing to `, as assual, via TCP` better?

and/or a socket (`unix:///path/to/proxy.sock`).

- it will connect to the API server and watch resources, like the current proxy;
- then, it will process them, applying Kubernetes specific business logic like topology computation relative to the local host;
- finally, provide the result of this computation to client via a gRPC watchable API.

> The idea is to sent the full state to the client, so implementations don't have to do diff-processing and maintain any internal state.

Reviewer: ```suggestion
The idea is to send the full state to the client, so implementations don't have to do diff-processing and maintain any internal state.
```

Reviewer: This assumes the client _wants_ the full state. But the userspace proxy can't do the "overwrite the entire old state with the new state" trick like iptables and ipvs do; it needs diffs. If the daemon only provides full states, then the userspace v2 implementation would need to maintain a copy of the full state and implement diffs itself. So maybe the daemon should support either mode?

Author: I wondered the same, but (1) it is my understanding that any k8s controller has a resync loop anyway and (2) could not see any case where the implementation does not have the current state. For instance, with the userspace proxy, the current state will be a list or map of listeners, that can be compared to the expected state, and that is required to apply a diff anyway..

This should provide simple implementations, reliable results and still be quite optimal, since many kernel network-level objects are
updated via atomic replace APIs. It's also a protection from slow readers, since no stream has to be buffered.

Since the node-local state computed by the new proxy will be simpler and node-specific, it will only change when the result for the

> current node is actually changed. Since there's less data in the local state, change frequency is reduced. Testing on actual clusters

Reviewer: With some network plugins, there is no per-node kube-proxy equivalent, because Services are just a specific case of the general cluster-wide network programming that is handled by some other system. Eg, in the case of ovn-kubernetes, the single ovn-kubernetes master controller watches Service and Endpoint changes and then updates the OVN database to reflect those changes. The OVN DB changes are then propagated to the individual nodes as needed, where `ovn-controller` processes make the necessary routing/filtering/etc changes for that node.

So, ovn-kubernetes could benefit from a helper that processed some of the Service-related information for it, but it would only want a single helper that gave it a top-level view of Services, not a per-node helper that gave a node-level view...

Author: yes, it is expected that not every use-case will benefit from this, but will still benefit from the simpler and more stable API.

> showed a frequency reduction by 2 orders of magnitude.

Reviewer: frequency reduction of _what_?

Author: of events to react to; for 100 events (on Service, Node, etc), only 1 has been found to change the proxy's resulting state.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

- sending the full-state could be resource consuming on big clusters, but it should still be O(1) to the actual kernel definitions
  (the complexity of what the node has to handle cannot be reduced without losing functionality or correctness).

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

> A draft implementation exists [here](https://github.com/mcluseau/kube-proxy2/) and some performance testing has been done
> [here](https://github.com/mcluseau/kube-proxy2/blob/master/doc/proposal.md).

Reviewer: ```suggestion
A [draft implementation] exists and some [performance testing] has been done.

[draft implementation]: https://github.com/mcluseau/kube-proxy2/
[performance testing]: https://github.com/mcluseau/kube-proxy2/blob/master/doc/proposal.md
```

> The watchable API will be a long polling, taking a "last know state info" and returning a stream of objects. Proposed defition:

Reviewer: ```suggestion
The watchable API will be a long polling, taking a "last known state info" and returning a stream of objects. Proposed definition:
```

```proto
service Endpoints {
    // Returns all the endpoints for this node.
    rpc Next (NextFilter) returns (stream NextItem);
}

message NextFilter {
    // Unique instance ID to manage proxy restarts
    uint64 InstanceID = 1;
    // The latest revision we're aware of (0 at first)
    uint64 Rev = 2;
}

message NextItem {
    // Filter to use to get the next notification (first item in stream)
    NextFilter Next = 1;
    // A service endpoints item (any item after the first)
    ServiceEndpoints Endpoints = 2;
}
```

When the proxy starts, it will generate a random InstanceID, and have Rev at 0. So, a client (re)connecting with get the
new state either after a proxy restart or when an actual change occurs. The proxy will never send a partial state, only
full states. This means it waits to have all its Kubernetes watchers sync'ed before going to Rev 1.

The first NextItem in the stream will be the state info required for the next polling call, and any subsequent item will
be an actual state object. The stream is closed when the full state has been sent.

> The client library abstract those details away and provides the full state after each change, and includes a default Run

Reviewer: ```suggestion
The client library abstracts those details away and provides the full state after each change, and includes a default Run
```

      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.

* **What steps should be taken if SLOs are not being met to determine the problem?**

[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

> ## Alternatives

Reviewer: Particularly if you are providing a client library anyway, does there really have to be a daemon? Couldn't the shared Service+Endpoint processing code just go into the client library directly, and then it would create a goroutine to process raw events from the apiserver and provide cooked updates to the user?

I guess the separate daemon allows for implementing non-golang proxies, but is anyone doing that?

Author: if any implementation uses a library, then it has to be released with every change to this client. My suggestion allows to decouple the releases of implementations from the core, so they occur only when improving the implementation, or reacting to a change in the simplified model (that should be much less frequent than k8s business logic changes or fixes).

Reviewer 2: i added the Alternatives section just now, i guess we can resolve this and discuss it in there
